### PR TITLE
Simplify copying starter assets on remix

### PR DESCRIPTION
**Pull request by bencodeorg**
When copying starter assets on remix, the level associated with the channel (ie, project) being copied is the template level, not the "derived" level. Simplifying our asset copying logic to reflect this. I got confused by this, so updating so hopefully the next person won't go through the same thing :)

## Links

Discovered as part of the investigation described in this Slack thread: https://codedotorg.slack.com/archives/C03DBDN67B7/p1666101508728619

## Testing story

Tested manually that Javalab levels where assets are defined on the level and on the template continued to work when remixed.
- on Javalab level: /s/allthethings/lessons/44/levels/4
- on Javalab template: /s/csa5-2022/lessons/13/levels/1
- on Applab level: /s/csd6-2022/lessons/14/levels/10 (images are broken when you share in Applab, but that's a result of [this bug](https://codedotorg.atlassian.net/browse/SL-278?jql=text%20~%20%22remix%20asset%22) -- assets are appearing in manage assets modal)
- on Applab template: /s/csd6-2022/lessons/14/levels/6

### CODE REVIEW

**Comment by bencodeorg** [comment]
@sanchitmalhotra126 could use your eyes on this in case it sparks any memories from working on this originally

File: dashboard/legacy/middleware/helpers/asset_bucket.rb [hunk 1]
```diff
@@ -43,12 +43,14 @@ def copy_level_starter_assets(src_channel, dest_channel)
...
+    # For a template-backed level, the level associated with the channel
+    # is the template level
     level = Level.cache_find(channel.level_id)
-    return unless level
+    return unless level&.starter_assets
 
     dest_owner_id, dest_storage_app_id = storage_decrypt_channel_id(dest_channel)
 
-    (level&.project_template_level&.starter_assets || level.starter_assets || []).map do |friendly_name, uuid_name|
+    level.starter_assets.map do |friendly_name, uuid_name|
```
Comment (molly-moen) [inline dashboard/legacy/middleware/helpers/asset_bucket.rb:56]: can you add a comment here explaining why we can always just copy the assets on the level?

Comment (bencodeorg) [reply dashboard/legacy/middleware/helpers/asset_bucket.rb:56]: updated, LMK what you think!

**Review by sanchitmalhotra126 — COMMENTED** [review]
Code looks good to me - can we just confirm this on App Lab as well? App Lab now also uses this code path after we added it for Java Lab

Edit: oh also, we use similar logic in [JavalabFilesHelper](https://github.com/code-dot-org/code-dot-org/blob/019b2f631c5521ec5b46def63c599671dd62c649/dashboard/app/helpers/javalab_files_helper.rb#L102) when copying starter assets. Trying to figure out if we would need to update that too...not sure though since these levels do run fine once the assets are copied right?

**Comment by bencodeorg** [comment]
> Code looks good to me - can we just confirm this on App Lab as well? App Lab now also uses this code path after we added it for Java Lab
> 
> Edit: oh also, we use similar logic in [JavalabFilesHelper](https://github.com/code-dot-org/code-dot-org/blob/019b2f631c5521ec5b46def63c599671dd62c649/dashboard/app/helpers/javalab_files_helper.rb#L102) when copying starter assets. Trying to figure out if we would need to update that too...not sure though since these levels do run fine once the assets are copied right?

Good call, will check applab.

Re: files helper, I'll see if this simplification applies there as well, but to clarify, this isn't a fix for something that's broken. Just a code simplification b/c we never hit the first condition (ie, looking for a level's template's starter assets)

**Comment by bencodeorg** [comment]
@sanchitmalhotra126 in the javalab_files_helper case, it looks like we pass in the level ID from the client, which looks like it reflects the derived level ID (just looked at editing an exemplar via levelbuilder), not the template level ID. So the full set of logic is useful there, I think.

In the remix case, we're deriving the level ID from the channel ID, which gives us the template's level ID. We could maybe refactor javalab_files_helper to also derive the level ID from the channel ID for consistency? Might be some perils there though

**Comment by sanchitmalhotra126** [comment]
> @sanchitmalhotra126 in the javalab_files_helper case, it looks like we pass in the level ID from the client, which looks like it reflects the derived level ID (just looked at editing an exemplar via levelbuilder), not the template level ID. So the full set of logic is useful there, I think.
> 
> In the remix case, we're deriving the level ID from the channel ID, which gives us the template's level ID. We could maybe refactor javalab_files_helper to also derive the level ID from the channel ID for consistency? Might be some perils there though

Gotcha, thanks for checking on that. I think leaving it as is is fine in that case

**Comment by bencodeorg** [comment]
@sanchitmalhotra126 checked out Applab levels, looks good

**Review by molly-moen — APPROVED** [review]
LGTM!

**Review by fisher-alice — COMMENTED** [review]
Thanks for simplifying the code here, Ben!

Considering the [Slack discussion](https://codedotorg.slack.com/archives/C03DBDN67B7/p1667835012490549?thread_ts=1666101508.728619&cid=C03DBDN67B7) surrounding this PR, I wonder if we could also block the uploading of starter assets in a level that contains a project_template_level. 

Within `level_starter_assets_controller.rb`, we could assign starter_assets as 
`starter_assets = (@level&.project_template_level&.starter_assets || (!@level.project_template_level && @level.starter_assets) || [])` instead of `(@level&.project_template_level&.starter_assets ||  @level.starter_assets) || [])` which would effectively block the uploading of assets on a level that contains a project_template_level so a curriculum-writer would only be able to add starter assets to the project_template_level. 

Let me know what you think. Thanks!

File: dashboard/legacy/middleware/helpers/asset_bucket.rb (hunk 1 above)
Comment (fisher-alice) [reply dashboard/legacy/middleware/helpers/asset_bucket.rb:56]: Thanks for adding the clarifying comment! [This file](https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/app/models/parent_levels_child_level.rb#L68) helped me understand the relationship between 'derived' or 'child' level and the project_template_level.

**Comment by bencodeorg** [comment]
> Thanks for simplifying the code here, Ben!
> 
> Considering the [Slack discussion](https://codedotorg.slack.com/archives/C03DBDN67B7/p1667835012490549?thread_ts=1666101508.728619&cid=C03DBDN67B7) surrounding this PR, I wonder if we could also block the uploading of starter assets in a level that contains a project_template_level.
> 
> Within `level_starter_assets_controller.rb`, we could assign starter_assets as `starter_assets = (@level&.project_template_level&.starter_assets || (!@level.project_template_level && @level.starter_assets) || [])` instead of `(@level&.project_template_level&.starter_assets || @level.starter_assets) || [])` which would effectively block the uploading of assets on a level that contains a project_template_level so a curriculum-writer would only be able to add starter assets to the project_template_level.
> 
> Let me know what you think. Thanks!

Definitely agree we should block uploading starter assets on child/derived levels on levelbuilder. That said, I would lean towards making controller updates in conjunction with a UI change that tells levelbuilders why they can't upload assets on a given level. I added a Jira item to track that work here:

https://codedotorg.atlassian.net/browse/SL-368

**Review by sanchitmalhotra126 — APPROVED** [review]

**Comment by fisher-alice** [comment]
> 

That definitely makes sense - Thanks!

**Pull request merged**